Sources/AsyncHTTPClient/ConnectionPool/HTTPRequestStateMachine.swift

```diff
@@ -0,0 +1,625 @@
+            return self.startSendingRequestHead(head)
+
+        case .running(.streaming(_, _, producer: .paused), .receivingBody(let head, _)) where head.status.code >= 300:
+            // If we are receiving a response with a status of >= 300, we should not send out
+            // further request body parts. The remote already signaled with status >= 300 that it
+            // won't be interested. Let's save some bandwidth.
+            return .wait
+
+        case .running(.streaming(let expectedBody, let sentBodyBytes, producer: .paused), let responseState):
+            let requestState: RequestState = .streaming(
+                expectedBodyLength: expectedBody,
+                sentBodyBytes: sentBodyBytes,
+                producer: .producing
+            )
+
+            self.state = .running(requestState, responseState)
+            return .resumeRequestBodyStream
+        }
+    }
+
+    private mutating func channelIsNotWritable() -> Action {
+        self.isChannelWritable = false
+
+        switch self.state {
+        case .initialized,
+             .verifyingRequest,
+             .waitForChannelToBecomeWritable,
+             .running(.streaming(_, _, producer: .paused), _),
+             .running(.endSent, _),
+             .finished,
+             .failed:
+            return .wait
+
+        case .running(.streaming(let expectedBodyLength, let sentBodyBytes, producer: .producing), let responseState):
+            let requestState: RequestState = .streaming(
+                expectedBodyLength: expectedBodyLength,
+                sentBodyBytes: sentBodyBytes,
+                producer: .paused
+            )
+            self.state = .running(requestState, responseState)
+            return .pauseRequestBodyStream
+        }
+    }
+
+    mutating func readEventCaught() -> Action {
+        switch self.state {
+        case .initialized,
+             .verifyingRequest,
+             .waitForChannelToBecomeWritable,
+             .running(_, .waitingForHead),
+             .running(_, .endReceived),
+             .finished,
+             .failed:
+            // If we are not in the middle of streaming the response body, we always want to get
+            // more data...
+            return .read
+        case .running(_, .receivingBody(_, .downstreamIsConsuming(readPending: true))):
+            preconditionFailure("It should not be possible to receive two reads after each other, if the first one hasn't been forwarded.")
+        case .running(let requestState, .receivingBody(let responseHead, .downstreamIsConsuming(readPending: false))):
+            self.state = .running(requestState, .receivingBody(responseHead, .downstreamIsConsuming(readPending: true)))
+            return .wait
+        case .running(let requestState, .receivingBody(let responseHead, .downstreamHasDemand)):
+            self.state = .running(requestState, .receivingBody(responseHead, .downstreamHasDemand))
```

Comment:
There's no state change here. Is that intentional?

Reply:
Yes.

Reply:
Yes. Added a comment.